File.delete async

pkar70 opened the pull request:
GitHub Issue (If applicable): #1187 (partly)

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
StorageFile:DeleteAsync not supported (workaround exist)

## What is the new behavior?
StorageFile:DeleteAsync is supported.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [X] Contains **NO** breaking changes
- [X] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [X] Associated with an issue (GitHub or internal)

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->


## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

jeromelaban commented on src/Uno.UWP/Storage/StorageFile.cs:
```diff
@@ -34,6 +34,18 @@ private StorageFile(Uri uri)
 			_fileUri = uri;
 		}
 
+		public async Task DeleteAsync()
```
It's best if you invoke `DeleteAsync(CancellationToken.None)`. Also, to avoid the missing API issue you'll need to keep the `IAsyncAction` return type. Use `AsyncAction.FromTask()` to return one.

pkar70 replied:
When I tried to
IAsyncAction DeleteAsync() => DeleteAsync(CancellationToken.None);
it complains that Task is not convertible to IAsyncAction.
Yes, in Generated, we have IAsyncAction DeleteAsync(), but in API (and in DeleteAsync(CancellationToken ct)) we have Task.
So, maybe more appropriate is adding method to IgnoreSet?

pkar70 replied:
And, why build complains about missing
Windows.Foundation.IAsyncAction Windows.Storage.StorageFolder.DeleteAsync()
as in StorageFolder.cs, lines 217-221, method
public IAsyncAction DeleteAsync() =>
is defined?

jeromelaban commented on src/Uno.UI.RuntimeTests/Tests/Windows_Storage/Given_StorageFile.cs:
```diff
@@ -0,0 +1,45 @@
+		[TestMethod]
+		public async void When_DeleteFile()
+		{
+      var folder = Windows.Storage.ApplicationData.Current.LocalFolder;
```
Indentation is off for the whole method.

pkar70 commented:
Build complains that `Windows.Foundation.IAsyncAction Windows.Storage.BulkAccess.FileInformation.DeleteAsync()` is not implemented. But why?
In `src/Uno.UWP/Generated/3.0.0.0/Windows.Storage.BulkAccess/FileInformation.cs`, DeleteAsync is still defined (as not implemented).

And, as it seems, many Windows.Storage.BulkAccess.FileInformation attributes/methods are defined in Windows.Storage.StorageFile, so it can be redirected (i.e. =>). But what class is more appropriate for method definition, and what class - for only redirection?
Windows.Storage.BulkAccess.DeleteAsync() => Windows.Storage.StorageFile.DeleteAsync() , or
Windows.Storage.StorageFile.DeleteAsync() => Windows.Storage.BulkAccess.DeleteAsync() ?

jeromelaban commented:
You've removed the DeleteAsyncMethod from the interface completely. You'll need to create a partial class for `IStorageItem` and add the method there.

pkar70 commented:
Three questions.
First, about CI complains that Windows.Foundation.IAsyncAction Windows.Storage.BulkAccess.FileInformation.DeleteAsync() is not implemented. 

"You'll need to create a partial class for IStorageItem and add the method there" - but what IStorageItem has to do with existing definition of BulkAccess.FileInformation.DeleteAsync()? Same function has to be both in IStorageItem and BulkAccess.FileInformation.DeleteAsync()?


Second.
Same methods/attributes exist both in StorageFile and BulkAccess.FileInformation. It seems convenient to implement such methods only once, and call them from second place. My question: in which class such metod should be implemented? In StorageFile or in BulkAccess.FileInformation ?

Third.
CI complains that Windows.Foundation.IAsyncAction Windows.Storage.StorageFolder.DeleteAsync()  is removed. But exactly this method is implemented (in  StorageFolder.cs, lines 217-221). So, as far as I can see, it is implemented, and not removed. What is wrong?

jeromelaban commented:
> Three questions.
> First, about CI complains that Windows.Foundation.IAsyncAction Windows.Storage.BulkAccess.FileInformation.DeleteAsync() is not implemented.
> 
> "You'll need to create a partial class for IStorageItem and add the method there" - but what IStorageItem has to do with existing definition of BulkAccess.FileInformation.DeleteAsync()? Same function has to be both in IStorageItem and BulkAccess.FileInformation.DeleteAsync()?
> 

It does, the [`FileInformation` class implements `IStorageFile`](https://github.com/unoplatform/uno/blob/6f7e23f71084cc7e785fbd0838d04eeeffe7f5e9/src/Uno.UWP/Generated/3.0.0.0/Windows.Storage.BulkAccess/FileInformation.cs#L8).

> Second.
> Same methods/attributes exist both in StorageFile and BulkAccess.FileInformation. It seems convenient to implement such methods only once, and call them from second place. My question: in which class such metod should be implemented? In StorageFile or in BulkAccess.FileInformation ?
> 

That's not how Microsoft implemented it, or they have some common internal class to provide the same feature for both. For now implementing `StorageFile` seems to be enough, let's leave `FileInformation` not implemented.

> Third.
> CI complains that Windows.Foundation.IAsyncAction Windows.Storage.StorageFolder.DeleteAsync() is removed. But exactly this method is implemented (in StorageFolder.cs, lines 217-221). So, as far as I can see, it is implemented, and not removed. What is wrong?

The signature has changed, [see my comment above](https://github.com/unoplatform/uno/pull/2407#discussion_r361825138) about returning `IAsyncAction`, not a `Task`.

pkar70 commented:
1) Should  `StorageFile.DeleteAsync(CancellationToken ct)` be converted from Task to IAsyncAction, to be 'in pair' with `StorageFile.DeleteAsync()`?
2) "removed Windows.Foundation.IAsyncAction Windows.Storage.StorageFolder.DeleteAsync()", and StorageFolder.cs, line 215: "public IAsyncAction DeleteAsync()". Both has IAsyncAction...
3) if I change StorageFile.DeleteAsync from Task to IAsyncAction, would messages about missing `Windows.Storage.BulkAccess.FileInformation.DeleteAsync()` disappear? Probably not?
4) I cannot find any Windows.Storage.BulkAccess definition (only in Generated), did I overlooked something?

jeromelaban commented:
> 1. Should  `StorageFile.DeleteAsync(CancellationToken ct)` be converted from Task to IAsyncAction, to be 'in pair' with `StorageFile.DeleteAsync()`?

No, as this method does not exist in WinRT.

> 2. "removed Windows.Foundation.IAsyncAction Windows.Storage.StorageFolder.DeleteAsync()", and StorageFolder.cs, line 215: "public IAsyncAction DeleteAsync()". Both has IAsyncAction...
> 3. if I change StorageFile.DeleteAsync from Task to IAsyncAction, would messages about missing `Windows.Storage.BulkAccess.FileInformation.DeleteAsync()` disappear? Probably not?
> 4. I cannot find any Windows.Storage.BulkAccess definition (only in Generated), did I overlooked something?

You removed `IStorageFile.DeleteAsync()`: https://github.com/unoplatform/uno/pull/2407/files#diff-70fd5e3931b53c4300ed9a08be7e957dR40

It's not supposed to be removed, unless the definition is placed in another file.

pkar70 commented:
Yes, I removed `IStorageFile.DeleteAsync()`.
So, I had to implement it / add it to:
Windows.Storage.IStorageItem.DeleteAsync() not found in ignore set.
Windows.Storage.StorageFile.DeleteAsync() not found in ignore set.
Windows.Storage.StorageFolder.DeleteAsync() not found in ignore set.
Windows.Storage.BulkAccess.FileInformation.DeleteAsync() not found in ignore set.
Windows.Storage.BulkAccess.FolderInformation.DeleteAsync() not found in ignore set.

I added it only to StorageFile (with Task, not IAsyncAction). So, I need to add it to all other. E.g. to StorageFolder. But, as it seems, it is already defined in StorageFolder (line 215).

Where I should place implementation of DeleteAsync() inside BulkAccess.FileInformation and BulkAccess.FolderInformation? Should I create such files, with only one method?

jeromelaban commented:
> Yes, I removed `IStorageFile.DeleteAsync()`.

You must not remove it from the `IStorageFile` interface, this is a breaking change. This is not a required change, this interface is rarely used anyways, so it's best if you keep it untouched for now.

It is that single change that creates the other missing methods in the other types.

pkar70 commented:
If I remember correctly, I had to remove it from IStorageFile, because build complains that StorageFile.DeleteAsync() is redefinition of IStorageFile.DeleteAsync() .

jeromelaban commented on src/Uno.UI.RuntimeTests/Tests/Windows_Storage/Given_StorageFile.cs:
```diff
@@ -0,0 +1,66 @@
+			}
+			catch
+			{
+			  Assert.IsNotNull(null, "CreateFile exception - error outside tested method");
```
`Assert.Fail`

pkar70 commented on src/Uno.UI.RuntimeTests/Tests/Windows_Storage/Given_StorageFile.cs:
```diff
@@ -0,0 +1,66 @@
+		[TestInitialize]
+		public void Init()
+		{
+			_filename = DateTime.Now.ToString("yyyyMMddHHmmss") + ".txt";
```
I thought that all tests has 'private' filesystem, so files from one test doesn't interfere with files from other tests. Isn't it correct?

carldebilly commented on src/Uno.UI.RuntimeTests/Tests/Windows_Storage/Given_StorageFile.cs:
```diff
@@ -0,0 +1,66 @@
+			}
+
+		  // try delete file
+			try
```
Please try to align comments with code ;-)
